File: benchmarks/mitata/encode_to_hex.js
import { run, bench, group, summary } from "mitata";

import { encodeToHex } from "llrt:codec";

function hyblidToHex(input) {
  return input instanceof Uint8Array ? input.toHex() : encodeToHex(input);
}

const input1 = Buffer.from("hello world");
console.log(Buffer.from(input1).toString("hex"));
console.log(encodeToHex(input1));
console.log(input1.toHex());

[
  ["small", Buffer.from("hello world")],
  [
    "middle",
    Buffer.from(
      "hello world this is a longer buffer example for iteration testing"
    ),
  ],
  [
    "large",
    Buffer.from(
      "hello world this is a longer buffer example for iteration testing, hello world this is a longer buffer example for iteration testing"
    ),
  ],
].forEach(([type, input]) => {
  group(type, () => {
    summary(() => {
      // bench("Buffer.from().toString()", () => {
      //   Buffer.from(input).toString("hex");
      // });
      bench("llrt:codec.encodeToHex()", () => {
        encodeToHex(input);
      });
      bench("Uint8Array.toHex()", () => {
        input.toHex();
      });
    });
  });
});

const input2 = "hello world";
console.log(Buffer.from(input2).toString("hex"));
console.log(encodeToHex(input2));

[
  ["small", "hello world"],
  [
    "middle",
    "hello world this is a longer buffer example for iteration testing",
  ],
  [
    "large",
    "hello world this is a longer buffer example for iteration testing, hello world this is a longer buffer example for iteration testing",
  ],
].forEach(([type, input]) => {
  group(type, () => {
    summary(() => {
      bench("Buffer.from().toString()", () => {
        Buffer.from(input).toString("hex");
      });
      bench("llrt:codec.encodeToHex()", () => {
        encodeToHex(input);
      });
    });
  });
});

await run();

File: libs/llrt_encoding/benches/encode_to_hex.rs
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
use criterion::{criterion_group, criterion_main, Criterion};
use llrt_encoding::*;
use rquickjs::{Context, Runtime};

fn criterion_benchmark(c: &mut Criterion) {
    let seed = "hello world this is a longer buffer example for iteration testing, hello world this is a longer buffer example for iteration testing";

    let runtime = Runtime::new().unwrap();
    runtime.set_max_stack_size(512 * 1024);

    let ctx = Context::full(&runtime).unwrap();

    let _ = ctx.with(|ctx| ctx.eval::<String, &[u8]>(b"import {encodeToHex} from 'llrt:codec';"));

    let func = ["const seed = Buffer.from('", seed, "';"].concat();
    let _ = ctx.with(|ctx| ctx.eval::<String, &[u8]>(func.as_bytes()));

    c.bench_function("[JS] Buffer::from(bytes).toString('hex')", |b| {
        let func = ["Buffer.from('", seed, "').toString('hex');"].concat();
        ctx.with(|ctx| b.iter(|| ctx.eval::<String, &[u8]>(func.as_bytes())));
    });

    c.bench_function("[JS] Uint8Array.toHex()", |b| {
        ctx.with(|ctx| b.iter(|| ctx.eval::<String, &[u8]>(b"seed.toHex()")));
    });

    c.bench_function("[JS] llrt:codec.encodeToHex()", |b| {
        ctx.with(|ctx| b.iter(|| ctx.eval::<String, &[u8]>(b"encodeToHex(seed)")));
    });

    c.bench_function("[Native] llrt_encoding::bytes_to_hex()", |b| {
        b.iter(|| bytes_to_hex(seed.as_bytes()))
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
